common/raw_string_ostream.h

```diff
@@ -0,0 +1,69 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_RAW_STRING_OSTREAM_H_
+#define CARBON_COMMON_RAW_STRING_OSTREAM_H_
+
+#include "common/check.h"
+#include "common/ostream.h"
+
+namespace Carbon {
+
+// Implements streaming output with an underlying string. The string must always
+// be taken prior to destruction.
+//
+// Versus `llvm::raw_string_ostream`:
+// - Owns the underlying string.
+// - Supports `pwrite` for compatibility with more stream uses in tests.
+class RawStringOstream : public llvm::raw_pwrite_stream {
+ public:
+  explicit RawStringOstream() : llvm::raw_pwrite_stream(/*Unbuffered=*/true) {}
+
+  // Move constructor.
+  RawStringOstream(RawStringOstream&& other) noexcept {
+    str_ = std::move(other.str_);
```

Comment:
```suggestion
    str_ = std::exchange(other.str_, "");
```

Otherwise I think if the string happens to be small enough for SSO, the CHECK in the destructor will fire unless both `*this` and `other` are taken.

Alternatively, does this need to be movable at all?

Reply:
Removed. I think it'd been a compiler failure at one point in the change, in particular I think with ErrorBuilder, but other edits seem to have smoothed it over.

Reply:
Nevermind, I just wasn't building the right thing.
https://github.com/carbon-language/carbon-lang/actions/runs/12835060089/job/35793679613

Reply:
Here's an approach that uses a pointer, if you prefer. Maybe better anyways?

Reply:
Yeah, I like this better.